core/src/main/java/io/questdb/cairo/CairoEngine.java

```diff
@@ -389,97 +532,128 @@ public long getUnpublishedWalTxnCount() {
             int segmentId,
             long walRowCount
     ) {
-        if (tableSequencerAPI.hasSequencer(tableName)) {
-            // This is WAL table because sequencer exists
-            return new WalReader(configuration, tableName, walName, segmentId, walRowCount);
+        if (tableToken.isWal()) {
+            return new WalReader(configuration, tableToken, walName, segmentId, walRowCount);
         }
 
-        throw CairoException.nonCritical().put("WAL reader is not supported for table ").put(tableName);
+        throw CairoException.nonCritical().put("WAL reader is not supported for table ").put(tableToken);
     }
 
-    @Override
-    public @NotNull WalWriter getWalWriter(CairoSecurityContext securityContext, CharSequence tableName) {
+    @TestOnly
+    public @NotNull WalWriter getWalWriter(CairoSecurityContext securityContext, TableToken tableToken) {
         securityContext.checkWritePermission();
-        return walWriterPool.get(tableName);
+        verifyTableToken(tableToken);
+        return walWriterPool.get(tableToken);
     }
 
     public TableWriter getWriter(
             CairoSecurityContext securityContext,
-            CharSequence tableName,
+            TableToken tableToken,
             String lockReason
     ) {
         securityContext.checkWritePermission();
-        validNameOrThrow(tableName);
-        return writerPool.get(tableName, lockReason);
+        verifyTableToken(tableToken);
+        return writerPool.get(tableToken, lockReason);
     }
 
     public TableWriter getWriterOrPublishCommand(
             CairoSecurityContext securityContext,
-            CharSequence tableName,
+            TableToken tableToken,
             @NotNull AsyncWriterCommand asyncWriterCommand
     ) {
         securityContext.checkWritePermission();
-        validNameOrThrow(tableName);
-        return writerPool.getWriterOrPublishCommand(tableName, asyncWriterCommand.getCommandName(), asyncWriterCommand);
+        verifyTableToken(tableToken);
+        return writerPool.getWriterOrPublishCommand(tableToken, asyncWriterCommand.getCommandName(), asyncWriterCommand);
+    }
+
+    public TableWriter getWriterUnsafe(TableToken tableToken, String lockReason) {
+        return writerPool.get(tableToken, lockReason);
+    }
+
+    public boolean isTableDropped(TableToken tableToken) {
+        return tableNameRegistry.isTableDropped(tableToken);
+    }
+
+    public boolean isWalTable(TableToken tableToken) {
+        return tableToken.isWal();
     }
 
     public String lock(
             CairoSecurityContext securityContext,
-            CharSequence tableName,
+            TableToken tableToken,
             String lockReason
     ) {
         assert null != lockReason;
         securityContext.checkWritePermission();
 
-        validNameOrThrow(tableName);
-        String lockedReason = writerPool.lock(tableName, lockReason);
+        String lockedReason = writerPool.lock(tableToken, lockReason);
```

Comment:
> Such code might lead to deadlocks.

This part is wrong, my bad. Unlock order as such doesn't lead to deadlocks - it only matters if a thread tries to acquire a lock when it released another one. Locking order is what creates dependency cycles leading to deadlocks. But unlocking in the reverse order is a common practice since it minimizes the time when other threads aren't able to proceed. Marked this comment as nit.

Reply:
changed in [1abaa43](https://github.com/questdb/questdb/pull/2752/commits/1abaa431fd00b2ab1adc0d87563795ce33f228a1)